### ISSUE TITLE
Convenient sudo() wrapper

### ISSUE THREAD

**Issue by bitprophet** [pyinvoke/invoke#294]
tl;dr Fabric 1's `sudo()` except implemented on top of generic, user-exposed features (#289, #293) and available at Invoke-on-up, since it's not actually specific to remote execution.

Given those user-exposed features, there's no technical necessity to implement it in core, because it *should* just turn into something like this:

``` python
@task
def identity(c):
    assert c.run("whoami").stdout.strip() == "myuser"
    sudo_prompt = "Sudo password plz:"
    with c.prefix("sudo -S -p '{0}' ".format(sudo_prompt):
        result = c.run("whoami", responses={sudo_prompt: "mysudopassword\n"})
        assert result.stdout.strip() == "root"
```

But clearly, even if you remove the asserts, that's annoying boilerplate we can erase by implementing a commonly used pattern as a convenience method.

----

MUCH LATER EDIT: There *is* a wrinkle here - in Fabric 1, this functionality includes "prompt the user ourselves when the sudo-prompt response value isn't pre-set, then cache the result from then on". That prompt+cache happens "inline" within the `sudo()` call. Brainstorm:

* The most-obvious answer is "you should store it in your config up front so no prompting is necessary".
* But storing that data on-disk is insecure; a runtime prompt is better.
    * A runtime "pulled from a secure store over the network, flash key, etc" is best. No prompt, no on-disk.
* But there's nothing saying we can't explicitly prompt up-front in `sudo`.
* Prompting "lazily" / as late as possible, is "nice" insofar as it means users with passwordless sudo never have to deal with a prompt at all - no inner prompt seen, no outer prompt displayed.
* In *most* cases, that is still just "nice" and not *required* - users who know up front that they will need the password, can simply enter it up front.
    * But the Fabric-specific use case of remote sudo is more complex - users with large fleets may not know  whether the prompt will ever appear.
    * Counterpoint: users with large fleets will usually want "raise/store an exception and keep trucking" instead...
* Users may also have heterogenous sudo passwords across their fleets, so a single up-front prompt is no longer enough - it'd need to be one prompt per host/group/whatever.
    * Counterpoint, if you have that setup, you could be using Fabric 2's per-host/group configs
    * But then you're back to storing the data on-disk
    * So you really *do* need one prompt per remote password, no matter when it pops up.
    * If you know ahead of time *where* the passwords differ, we could then say "ok, you just need to prompt for each group" (perhaps with a config setting like e.g. `{'hostname': {'prompt_for_password': True}}`)
    * If you *don't*, that is when runtime inline prompting becomes truly "necessary". Whether that's enough of a use case to be worth implementing this ourselves, hard to say.
    * **NOTE** that all of these concerns pop up with Fabric connection/login passwords too - with the exception that the point of prompting is at connection time and not anywhere in `Runner`.
* One way to enable "late but not inline" prompting is to have the prompt machinery raise an exception when it encounters a prompt key whose value is `None`, then `sudo()` can treat this as "prompt, fill config, try that same command again"
    * This isn't **exactly** the same as Fabric 1's behavior, because it involves running the command literally twice in a row.
    * But it feels cleaner because it doesn't require the truly-inline prompting.
    * It also arguably lets us push the entirety of "dealing with missing password caches" on the user, which is better for 'training' them to do the right thing, but worse from a usability standpoint.
* If we *did* go with fully inline prompting again, at least we can make it more generic - again, anything whose value is `None` could trigger the pause-and-prompt, not just sudo specific passwords.
    * Though there's the added wrinkle of "when to use `getpass.getpass` vs `(raw_)input`"...
* Yet another reason *not* to do inline prompting is that it adds implementation complexity - e.g. in Fabric 1 we need the prompt machinery to talk to the stdin handler so the stdin handler pauses itself while prompting is happening (otherwise `getpass` or `raw_input` never see the stdin, as it's being intercepted).

**Comment by bitprophet** [comment]
In the interests of getting the fab 2 alpha out the door, I think my initial target for this feature, re: sudo password prompting, will be "it requires the autoresponse password value to be non-`None`". Users configuring it - as many do (despite it being technically insecure) in Fabric 1 via `~/.fabricrc` + `sudo_password = xxx` - never have to see a prompt; it will prompt prior to execution if the value has been left unset.

We can revisit whether to add the "prompt lazily to fill the password, then rerun the command" and/or "actually prompt inline during autoresponding" aspects of the functionality, later.

**Comment by bitprophet** [comment]
Another minor fun wrinkle that just occurred to me: Fabric 2 rebinds Invoke's `run` as `local`...what should it do with Invoke's `sudo`? Grump.

**Comment by bitprophet** [comment]
Occurred to me recently: the old "run a remote-oriented script locally instead" use case for Fabric (fabric/fabric#98) *is* one major reason to make `sudo` implemented in Invoke proper. (Also had a user asking about it on IRC recently too...)

The problem of how to expose the API in both libs remains open, none of the solutions I can think of are awesome unfortunately:

* Don't rebind, just retarget: `Context.sudo` executes local sudo, `Remote.sudo` executes remote sudo, and there *is no* `Remote.local_sudo` or similar.
    * "Solves" the API collision problem, and leans on the part where I sincerely doubt a task targeting remote servers would need to do anything superuser-ly locally. And is of no consequence in the fab-98 situation - sudo is sudo is sudo.
    * But is inconsistent with run-vs-local, and perhaps unnecessarily annoys anyone who does inexplicably need to sudo on both ends.
* Just rebind it, you doofus: `Context.sudo` becomes `Remote.local_sudo`.
    * Inelegant, sure, but does what it says on the tin, and because it's barely going to be used by anybody, who really cares? It's like 3 lines in the API doc.
* Somehow turn "sudo-ing" into a kwarg to `run`/`local`, e.g. `Context.run('xxx', sudo=True)`.
    * Not amazing as it pushes the sudo implementation (including all the phase-2 prompt related logic noted in ticket desc) inside `run` in some fashion, which at the very least means users tweaking the sudo implementation is now intertwined with any manipulation of `run`
    * OTOH, this stuff can all still happen at the `Context` level (i.e. would not actually impact `Runner`) so it could maybe be as simple as `return (self._sudo if sudo else self._run)(command, **kwargs).
    * But still feels like a lot of mental overhead for implementers and users just to avoid some renaming...
    * Also doesn't play nice with fabric/fabric#25, aka "how to select internal `sudo` for things like `put`/`get` and friends", since it removes `sudo` as a distinct top level execution target.

**Comment by bitprophet** [comment]
Moved this back to Invoke and continued actually implementing it. Mostly there, sticking-up bits that need addressing now instead of later:

* Thought I'd be clever by selecting an obviously-not-a-real-sudo-prompt string as the default forced password prompt, so that debugging would be more obvious. Unfortunately, if not paired with a Fabric 1 style "implicitly strip out that prompt from the mirrored+captured stderr stream", this looks pretty ugly.
    * Even if it was a regular-looking prompt, I assume it could still be pretty confusing to some users, especially if execution somehow got lagged or hung while the prompt was the last thing on the screen. Which is another argument for implementing "strip out things being responded to automatically".
    * But a counterargument is that magic is bad, and also that this would require more poking around inside `run` just to support `sudo` (though - if done it probably *should* be done for all auto-responses anyways).
    * For now I'll just revert to a "regular-looking" prompt, maybe with some text in it still making it clear it's part of the auto-response system.
* Integration testing this feels hard to do cleanly; even if we mock out `getpass`, there's the glaring issue of requiring all developers running integration tests, mutate their local system config so the tests can pass w/o human interaction.
    * Not only is that ugly and fragile, it's flat-out insecure as either this test-only user's creds need persisting in the public repo, or the user must store them in e.g. `~/.invoke.yml`, which is still putting a sudo-capable user's creds on disk.
    * I think the best middle ground is to make a selected-by-Travis-only integration test/suite, which of course *can* then rely on something always being present, *and* is ephemeral / not anybody's real system.

**Comment by bitprophet** [comment]
Made #365 for the hide-sudo-prompts-maybe angle.

**Comment by bitprophet** [comment]
And as noted in #365 too...without that in place, we get a silly thing where subsequent stdout starts on same line as the sudo prompt (which is stderr) as the autoresponded newline (if one's even present; it's not automatically added at this point) is not echoed to the local terminal.

I bet that'll come up in the alpha, but I'll wait until it does, for now. Not critical.

**Comment by bitprophet** [comment]
Reverted the sudo prompt to something more realistic looking (as implied above).

Integration test, grump, I forgot we'd switched to Travis' sudo-less (faster) architecture. Would be nice not to have to switch back, especially since they will presumably remove sudo-capable test runners entirely sometime. (OTOH, we hard require it for Fabric 1's suite so we can fix some frustrating and random-ass system level bugs, so same boat there if still applicable.)

Double checked their docs; maybe they will keep it around or we can work around it some other way, going by https://docs.travis-ci.com/user/ci-environment#Virtualization-environments and https://docs.travis-ci.com/user/trusty-ci-environment the 'Trusty' env is still being poked at / lives side by side with both legacy and containerized setups.

May well still be the case that tests are slower on Trusty than containerized, but if that's the price we have to pay, so be it?

For now, I'm going to punt on that, may make it another spinoff ticket.

**Comment by bitprophet** [comment]
Ran into another snag: I'd totally forgotten how Fabric 1 not only autoresponds with sudo password, but *also* detects the "Sorry, try again" output from `sudo` and uses that to determine whether it succeeded / whether to re-prompt / etc.

In a generic, Invoke/Fab 2 world, that requires #289's functionality to get beefed up sooner instead of later, so it can do more than simple always-respond-to-x-with-y. Will spin off another ticket for that soon after I have a brief think.

**Comment by bitprophet** [comment]
See #369 for thoughts on that. Think I see a way forwards that wouldn't even be that much code.

**Comment by bitprophet** [comment]
Aside from #369, which can help implement this, there's still the bigger question of exactly how to handle failed sudo passwords in a few ways - regardless of how exactly #369 lets us *detect* failure:

- How much to retry, if at all?
    - Could "only ever except", but that's not helpful to adhoc/interactive users.
    - Following Fab 1 still seems best: try forever until success or user KeyboardInterrupts, unless config/kwarg says to only-except.
    - Given the increased emphasis on "make this work for libraries/correctness first, interactive users second, when feasible" - maybe we make excepting the *default* behavior, and let users seriously wanting "full interactivity" turn on the "no, please prompt me if the password I gave was wrong/nonexistent" behavior?
- Should newly submitted passwords replace the old one the config (as in Fab 1)?
    - Not preserving the new password in some fashion feels bad, because if a user fat-fingered the first prompt, they shouldn't then have to re-submit the correct password for all subsequent `sudo()` calls.
        - So the question is, should an interactively prompted password be stored as a separate, not user-facing config path (e.g. `sudo.prompted_password`) or should it truly overwrite `sudo.password`?
    - We can only assume a single password (there's no concept of per-connection data as in Fabric 2) so I lean "yes" - keeping the original, clearly incorrect password around doesn't seem like it'd be useful. (Vs a Fabric 2 situation where the original password might be valid for other hosts later down the line.)
    - However, this sort of behind-the-scenes manipulation of data the user themselves presumably assembled, feels like a magical side effect, which is why "store it in a separate location" feels possibly better.

**Comment by bitprophet** [comment]
So given that, I think what this means is that the sudo version of `Responder` can simply be "detect failure, raise exception instead of responding again" (and thus could potentially be a generic, idk, `FailureDetectingResponder`).

Then `sudo` itself is responsible for noticing this exception & either reraising it or prompting, saving the new result as above, and trying again.

**Comment by bitprophet** [comment]
Have been mostly taking notes in TODO comments...bad habit.

At this point I have a working `FailureResponder` which takes a failure sentinel and raises `ResponseFailure` if it is seen. Still need to determine what `sudo` does when it sees one of these, as above.

On top of that: in the case(s) where we continue bubbling up the exception instead of retrying, it'd be nice for the exception to gain info about the `sudo` call it is bubbling up through...i.e. to turn into something like `Failure`/`Result`. Except that right now `Result` assumes it's a "finished" command, with attrs like `exit_code`, `ok`, etc - things which do not apply in a "sudo authentication failed partway" scenario.

So feels like what needs doing is:

- [ ] Rip out the "this is a regular command that failed during execution" language/assumptions from `Failure` to a new subclass called e.g. `CommandFailure`
- [ ] Create e.g. `AuthFailure`, subclassing `Failure`, for use by `sudo`
- [ ] Update `Context.sudo` so it transmutes `ResponseFailure` into `AuthFailure`
    - [ ] clients of `sudo` probably don't need to care about the details of the response failure itself, so it can just be discarded. Maybe someday we'll attach it to the new exception object, but that seems pretty pointless right now.
    - [ ] This *actually* becomes "update `run` so it lets callers such as `sudo` tell it how to deal with unexpected errors" - because we need the innards of `run`, which generate `Failure`/`Response`, to still do that, even when our `ResponseFailure` occurs. Unclear how best to handle this:
        - callbacks/injection, e.g. "here, call this with your `Response` when you see a `FooException`", hand it something that reraises the `AuthException` around the `Response`"
        - `run` learns to attach *any* inner exceptions to the raised `Failure` and then the caller is responsible for deciding whether to open and raise the inner exception, continue raising `Failure`, or etc
        - refactor `run` some more so the response-generating bits are torn out of the actually-running bits. (Not ideal since `run` is already split up into a few bits for various reasons.)
- [ ] Document clearly how auth failure interacts, or doesn't, with `warn=True` - rather, that `warn=True` only affects command failures, and not auth failures.

**Comment by bitprophet** [comment]
More ramblings about the watcher -> runner -> post-`run` exception handling BS:

- Out of the above, the 'best' scenario seemed to be "have Runner create `Failure` subclass instances for both command failure (normally) and watcher failure (by catching a Watcher-specific exception type)".
    - This avoids one tier of exceptions & exception handling - `sudo` no longer has to care at all, both subtypes of `Failure` will bubble up.
- Had another thought today of: why not pass the `Response` object in question (because `Failure` itself is mostly just a vehicle for the `Response` informing the catcher exactly what command is failing) into the watchers, allowing them to raise only one single exception that can bubble *all* the way up?
    - But that's probably tighter coupling than I want; it complicates Watcher and also spreads around the "how to turn a being-executed command request into a failure exception" logic to multiple classes.
    - Which doesn't seem worth it; I've managed to dilute my original overdone idea with 3 exception-handling steps, to 2, and going down to 1 isn't that much bigger a savings.

**Comment by bitprophet** [comment]
Pretty unhappy with how rabbit-hole-y this all got. Hope it pays off for somebody in the future somehow.

Still need to go thru some of the TODOs I spawned, and document the warn-vs-auth-failure bit. Then try this out via Fab 2 to make sure it works well that way.

**Comment by bitprophet** [comment]
My tolerance for which TODOs truly can't wait is now much lower, but there are still a few that should really be done now:

- [ ] Solve, or at least workaround/document, the problem where `Context.sudo` wants to submit its watcher to `run(..., watchers=[watcher])` but doesn't want to stomp on any actual user watchers.
    - Right now the implementation gives `watchers` as a kwarg, meaning user-modified config `watchers` will be overridden (kwarg always wins), and users passing in their own `watchers=` kwarg will encounter a "you gave the same kwarg twice!" error.
    - We can at least merge a user-supplied kwarg with our value, *but* that still leaves "user supplied config, did not supply kwarg, is confused when their config values are ignored because of *our* kwarg"
    - Feels sorta like the old "merging vs overriding" configuration cascade problem - which isn't truly solved for our config stuff yet either.
    - Maybe we just need to be fully explicit and *check* the config *and* kwarg and slot ours in alongside the highest-importance such value, if any are given. (And if none were, then we can just do what we do now and stick it in our own kwarg value.)
- [ ] AuthFailure handling at the `Program` level:
    - Currently not handled so stacktrace + implicit exit 1
    - Ideally would match `UnexpectedExitFailure` re: not showing stacktrace
    - Except right now, that actually does nothing but ensure the exit code matches; if you have stderr hidden, you'll see NO explanation of why it exited badly. In sudo's case, we definitely want to at least `sys.exit(str(e))` so our message is shown.
        - Probably also want to print more of our `Result` so the command generating the error is displayed...?
    - Ideally, `Program` should do as little as possible; I want these kinds of exits to look very similar between `inv` and eg `python -c "from invoke import blah; blah()"`, just with stacktraces hidden by default when you run via `Program`.
- [ ] There's some outstanding naming worries:
    - [ ] `s/Failure/Error` because that's Python/PEP8 style
    - [ ] stuff in the responder classes, some of the kwargs are a bit verbose

**Comment by bitprophet** [comment]
Minor wrinkle with the watcher kwarg/config merging - mutating a config object feels bad/wrong: if user introspects during debugging they'll wonder why the config does not actually reflect what they submitted; and we don't have a good story for merge-or-override in configs yet either.

Best solution for now is to rely on "kwarg beats config" by copying the config to be the kwarg, when only config exists. Then it resembles the base case of "user only gave kwarg", we append, and everyone's happy.

**Comment by bitprophet** [comment]
Punting the AuthFailure non-traceback stuff to #393.

**Comment by bitprophet** [comment]
Kinda wanted to `s/Failure/CommandFailure/g` but I can't reeeeeally justify the backwards compatibility hit. Maybe I'll do it just prior to 1.0. Otherwise tho, I dropped `Failure`-as-a-suffix from most other classes. (`AuthFailure` is purposefully semi-generic so it can stay.)

**Comment by Winand** [comment]
This sudo wrapper cannot prompt the user for password, right? Then I think one can put an object with the following method in `c.config.sudo.password` to prompt lazily in runtime:
```python
def __str__(self):
    if not self.value:
        desc = f" ({self.description})" if self.description else ""
        if self.hidden:
            # Prompt the user for a password without echoing.
            self.value = getpass.getpass(f"{self.name}{desc}: ")
        else:
            self.value = input(f"{self.name}{desc}: ")
    return self.value
```
